3 course/OperatingSystem/lab2/sync_list.c

```diff
@@ -0,0 +1,236 @@
+        sleep(1);
+        printf("-----------------------------------\n");
+        printf("Iter count less: %d \nIter count great: %d\nIter count equal: %d\nIter SWAP count: %d \n", iter_count_less, iter_count_great, iter_count_equal, iter_swap_count);
+
+    }
+}
+
+void generate_random_string(char *str, int min_length, int max_length) {
+    const char charset[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"; 
+    int len = rand() % (max_length - min_length + 1) + min_length; 
+
+    for (int i = 0; i < len - 1; ++i) {
+        int index = rand() % (strlen(charset) - 1); 
+        str[i] = charset[index];
+    }
+
+    str[len - 1] = '\0';
+}
+
+void append(Storage* list, char* data) {
+    Node* tmp = (Node*)malloc(sizeof(Node));
+    assert(tmp);
+
+    memcpy(tmp->value, data, strlen(data));
+    pthread_mutex_init(&tmp->sync, NULL);
+
+    Node* current_node = list->head;
+
+    if (current_node == NULL) {
+        list->head = tmp;
+        return;
+    } 
+    
+    while(current_node->next != NULL) {
+        current_node = current_node->next;
+    }
+    current_node->next = tmp;
+}
+
+Storage* init(int count) {
+    Storage* list = (Storage*)malloc(sizeof(Storage));
+    assert(list);
+
+    for (int i = 0; i < count; i++) {
+        char data[SIZE];
+        generate_random_string(data, 1, SIZE);
+        append(list, data);
+    }
+
+    return list;
+}
+
+//когда у нас 4 элемента, мы сначала захватываем лочку на первых трех элементах, затем, на 2-4
+//как только перекинули указатель first->next на третий элемент, захватываем лочку на 2-4 элементах
+void random_nodes_swap(Storage* list) {
+    assert(list->head);
+        
+    pthread_mutex_lock(&list->head->sync);
+    Node* first = list->head;
+    Node* second = NULL;
+    Node* third = NULL;
+
+    while (first->next) {
+        pthread_mutex_lock(&first->next->sync);
+        second = first->next;
+        if (second->next) {
+            pthread_mutex_lock(&second->next->sync);
+            third = second->next;
+            if (third != NULL && rand() % 2 == 1) { 
+                first->next = third;
+                second->next = third->next;
+                third->next = second;
+            }
+            pthread_mutex_unlock(&third->sync);
+        }
+        pthread_mutex_unlock(&first->sync);
+        first = second;
+    }
+    pthread_mutex_unlock(&first->sync);
+}
+
+void* swap_thread_func(void* arg) {
+    Storage* list = (Storage*)arg;
+
+    while (1) {
+        random_nodes_swap(list);
+        iter_swap_count++;
+    }
+
+    return NULL;
+}
+
+int less(size_t a, size_t b) {
+    return a < b;
+}
+
+int great(size_t a, size_t b) {
+    return a > b;
+}
+
+int equal(size_t a, size_t b) {
+    return a == b;
+}
+
+void count_good_pair(Storage* list, int (*cmp)(size_t, size_t)) {
+    if (list == NULL) {
+        return;
+    }
+
+    int cgp = 0; // считаем хорошие пары)))
+    Node* current = list->head;
+    while (current && current->next) {
```

Comment:
sorry, race here, you cannot use current->next because it might point to the end of the list while currect is in the begining